lib/device/manager.js

```diff
@@ -0,0 +1,325 @@
+ * Copyright (c) 2019, Nodari Chkuaselidze (MIT License)
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const Network = require('bcoin/lib/protocol/network');
+const Logger = require('blgr');
+const EventEmitter = require('events');
+const LedgerDeviceManager = require('./ledgermanager');
+const { vendors } = require('../common');
+
+const AVAILABLE_VENDORS = new Set(Object.values(vendors));
+const VENDOR_MANAGERS = {
+  [vendors.LEDGER]: LedgerDeviceManager
+};
+
+class DeviceManager extends EventEmitter {
+  constructor(options) {
+    super();
+
+    this.opened = false;
+    this.options = new DeviceManagerOptions(options);
+    this.network = this.options.network;
+    this.logger = this.options.logger;
+    this.vendorManagers = new Map();
+    this.cachedDevices = new Map();
+    this.selected = null;
+
+    this.init();
+  }
+
+  /**
+   * Initalize device manager,
+   * Create vendor managers.
+   * @private
+   */
+
+  init() {
+    for (const vendor of this.options.enabledVendors) {
+      const options = this.options.vendorManagerOptions.get(vendor);
+      const manager = new VENDOR_MANAGERS[vendor](options);
+
+      this.vendorManagers.set(vendor, manager);
+    }
+  }
+
+  /**
+   * Start listening to vendor manager events.
+   * @private
+   */
+
+  bind() {
+    const connect = (device) => {
+      const handle = getHandle(device);
+
+      assert(!this.cachedDevices.has(handle),
+        'Already have device for the handle.');
+
+      this.logger.debug('Device connected: ',
+        device.vendor,
+        device.key,
+        device.handle
+      );
+
+      this.cachedDevices.set(handle, device);
+      this.emit('connect', device);
+    };
+
+    const disconnect = (device) => {
+      const handle = getHandle(device);
+
+      this.logger.debug('Device disconnected: ',
+        device.vendor,
+        device.key,
+        device.handle
+      );
+      this.cachedDevices.delete(handle);
+      this.emit('disconnect', device);
+    };
+
+    const select = (device) => {
+      this.selected = device;
+      this.logger.debug('Device was selected: ',
+        device.vendor,
+        device.key,
+        device.handle
+      );
+      this.emit('select', device);
+    };
+
+    const deselect = (device) => {
+      if (device === this.selected) {
+        this.logger.debug('Device was deselected: ',
+          device.vendor,
+          device.key,
+          device.handle
+        );
+        this.selected = null;
+      }
+
+      this.emit('deselect', device);
+    };
+
+    for (const manager of this.vendorManagers.values()) {
+      manager.on('connect', connect);
+      manager.on('disconnect', disconnect);
+      manager.on('select', select);
+      manager.on('deselect', deselect);
+    }
+  }
+
+  /**
+   * Select device
+   * @param {String} vendor
+   * @param {DeviceWrapper} device
+   * @returns {DeviceWrapper}
+   */
+
+  async selectDevice(vendor, device) {
```

Comment:
To improve compatibility with other device managers:
 1. overload first parameter to accept STRING or Device type and
   act accordingly.
 2. Ignore first parameter if device is present and use vendor
   from the device (device.vendor)
 3. Check vendor === device.vendor and throw if they don't match
   but there are no benefits with this, you will just use
   selectDevice(device.vendor, device);